gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/control/RebalanceOperationImpl.java

```diff
@@ -156,29 +211,48 @@ public RebalanceResults getResults() throws CancellationException, InterruptedEx
           throw new InternalGemFireError(e.getCause());
         }
       }
+    }
+    return results;
   }
 
   public RebalanceResults getResults(long timeout, TimeUnit unit)
       throws CancellationException, TimeoutException, InterruptedException {
-    try {
-      return getFuture().get(timeout, unit);
-    } catch (ExecutionException e) {
-      if(e.getCause() instanceof GemFireException) {
-        throw (GemFireException) e.getCause();
-      } else if(e.getCause() instanceof InternalGemFireError) {
-        throw (InternalGemFireError) e.getCause();
-      } else {
-        throw new InternalGemFireError(e.getCause());
+    long endTime = unit.toNanos(timeout) + System.nanoTime();
+    
+    RebalanceResultsImpl results = new RebalanceResultsImpl();
+    List<Future<RebalanceResults>> frlist =  getFutureList();
+    for(Future<RebalanceResults> fr : frlist) {
+      try {
+        long waitTime = endTime - System.nanoTime();
+        RebalanceResults rr =  fr.get(waitTime, TimeUnit.NANOSECONDS);                
+        results.addDetails((RebalanceResultsImpl)rr);
+      } catch (ExecutionException e) {
+        if(e.getCause() instanceof GemFireException) {
+          throw (GemFireException) e.getCause();
+        } else if(e.getCause() instanceof InternalGemFireError) {
+          throw (InternalGemFireError) e.getCause();
+        } else {
+          throw new InternalGemFireError(e.getCause());
+        }
       }
```

Comment:
This else block looks a bit suspect. It should check for fatal errors (like OOME) instead of wrapping always in InternalGemFireError -- see comment before on how we handle fatal errors.